Composer/packages/client/src/pages/publish/Publish.tsx

```diff
@@ -3,357 +3,222 @@
+import { useState, useEffect, useMemo, Fragment, useRef } from 'react';
 import { RouteComponentProps } from '@reach/router';
 import formatMessage from 'format-message';
-import { Dialog } from 'office-ui-fabric-react/lib/Dialog';
-import { TextField } from 'office-ui-fabric-react/lib/TextField';
 import { useRecoilValue } from 'recoil';
-import { ActionButton } from 'office-ui-fabric-react/lib/Button';
-import { DialogSetting, PublishTarget, PublishResult } from '@bfc/shared';
-import isEqual from 'lodash/isEqual';
+import { PublishResult } from '@bfc/shared';
 
 import { dispatcherState, localBotPublishHistorySelector, localBotsDataSelector } from '../../recoilModel';
-import { Toolbar, IToolbarItem } from '../../components/Toolbar';
 import { AuthDialog } from '../../components/Auth/AuthDialog';
 import { createNotification } from '../../recoilModel/dispatchers/notification';
-import { Notification, PublishType } from '../../recoilModel/types';
+import { Notification } from '../../recoilModel/types';
 import { getSensitiveProperties } from '../../recoilModel/dispatchers/utils/project';
 import { armScopes } from '../../constants';
 import { getTokenFromCache, isShowAuthDialog, isGetTokenFromUser } from '../../utils/auth';
 import { AuthClient } from '../../utils/authClient';
 import TelemetryClient from '../../telemetry/TelemetryClient';
+import { ApiStatus, PublishStatusPollingUpdater, pollingUpdaterList } from '../../utils/publishStatusPollingUpdater';
 
 import { PublishDialog } from './PublishDialog';
 import { ContentHeaderStyle, HeaderText, ContentStyle, contentEditor } from './styles';
-import { BotStatusList, IBotStatus } from './BotStatusList';
+import { BotStatusList } from './BotStatusList';
 import { getPendingNotificationCardProps, getPublishedNotificationCardProps } from './Notifications';
 import { PullDialog } from './pullDialog';
+import { PublishToolbar } from './PublishToolbar';
+import { IBot, IBotStatus, IBotPublishHistory, IBotPublishType, IBotPublishTargets, IBotSetting } from './type';
 
-const publishStatusInterval = 10000;
 const deleteNotificationInterval = 5000;
-const intervals: { [key: string]: number } = {};
 
-const generateComputedData = (botProjectData, publishHistoryList, currentBotPublishTargetList) => {
-  const botSettingList: { projectId: string; setting: DialogSetting }[] = [];
-  const statusList: IBotStatus[] = [];
-  const botPublishTypesList: { projectId: string; publishTypes: PublishType[] }[] = [];
-  const publishTargetsList: { projectId: string; publishTargets: PublishTarget[] }[] = [];
+// CR: Data model. What is stable data? What is frequently changed data? Manage them separately rather than repeating them.
+const generateComputedData = (botProjectData) => {
+  const botSettingList: IBotSetting[] = [];
+  const botPublishTargetsList: IBotPublishTargets[] = [];
+  const botPublishTypesList: IBotPublishType[] = [];
+  const botList: IBot[] = [];
   botProjectData.forEach((bot) => {
```

Comment:
it looks like you can use

`botList= botProjectData.map(() =>` instead of forEach which makes u create a temporary variable

Reply:
Using `forEach` is because we need assign values to `botList`, `botSettingList`, `botPublishTargetsList` and `botPublishTypesList` in one loop

Reply:
You can do that inside map too.